compiler/hash-typecheck/src/ops/unify.rs

```diff
@@ -239,14 +240,74 @@ impl<'gs, 'ls, 'cd> Unifier<'gs, 'ls, 'cd> {
         // })
     }
 
-    // pub fn unify_args(&self, src_args: &Args, target_args: &Args) -> TcResult<TySub> {}
+    /// Convenience method to get a substitutor.
+    fn substitutor(&mut self) -> Substitutor {
+        Substitutor::new(self.storages_mut())
+    }
 
-    pub fn unify_tys(
-        &mut self,
-        src_id: TermId,
-        target_id: TermId,
-        opts: &UnifyTysOpts,
-    ) -> TcResult<Sub> {
+    /// Unify two substitutions to produce another substitution.
+    ///
+    /// The resultant substitution contains all the information of the two source substitutions,
+    /// without any common free variables in their domains.
+    ///
+    /// This implements the algorithm outlined in the paper:
+    /// <https://www.researchgate.net/publication/221600544_On_the_Unification_of_Substitutions_in_Type_Interfaces>
+    ///
+    pub fn unify_subs(&mut self, s0: &Sub, s1: &Sub) -> TcResult<Sub> {
+        let dom_s0: HashSet<_> = s0.domain().collect();
+        let dom_s1: HashSet<_> = s1.domain().collect();
+        let mut substitutor = self.substitutor();
+
+        /// First split the domains into three parts: d0, d1, and the intersection (see second
+        /// loop)
+        let d0: HashSet<_> = dom_s0.difference(&dom_s1).copied().collect();
+        let t0 = Sub::from_pairs(
+            d0.iter()
+                .map(|&a| (a, substitutor.apply_sub_to_subject(s0, a))),
+        );
+
+        let d1: HashSet<_> = dom_s1.difference(&dom_s0).copied().collect();
+        let t1 = Sub::from_pairs(
+            d1.iter()
+                .map(|&a| (a, substitutor.apply_sub_to_subject(s1, a))),
+        );
+
+        // Start with t0 and add terms for d1 one at a time, always producing well formed
+        // substitutions
+        let mut result = t0.clone();
+        for (a, t) in t0.pairs() {
+            // Remove elements of dom(result) from t, and remove a from result.
+            let subbed_t = substitutor.apply_sub_to_term(&result, t);
+            if substitutor.get_free_vars_in_term(subbed_t).contains(&a) {
+                panic!("Unexpected free variable in one of the substitutions being unified (occurs error)");
```

Comment:
In the future, maybe use reporting mechanism to outline the span of where this occurred

Reply:
Will add a note for the next PR.